dmoj/cptbox/ptdebug_arm64.cpp

```diff
@@ -0,0 +1,120 @@
+#define _BSD_SOURCE
+#include <sys/ptrace.h>
+#include <sys/uio.h>
+#include "ptbox.h"
+
+#define ARM_x0 0
+#define ARM_x1 1
+#define ARM_x2 2
+#define ARM_x3 3
+#define ARM_x4 4
+#define ARM_x5 5
+#define ARM_x6 6
+#define ARM_x7 7
+#define ARM_x8 8
+
+#ifdef PTBOX_NEED_PRE_POST_SYSCALL
+#include <elf.h>
+
+void pt_debugger_arm64::pre_syscall() {
+    struct iovec iovec;
+    iovec.iov_base = arm64_reg;
+    iovec.iov_len = sizeof arm64_reg;
+    if (ptrace(PTRACE_GETREGSET, tid, NT_PRSTATUS, &iovec))
+        perror("ptrace(PTRACE_GETREGSET)");
+
+    arm64_reg_changed = false;
+}
+
+void pt_debugger_arm64::post_syscall() {
+    if (!arm64_reg_changed)
+        return;
+
+    struct iovec iovec;
+    iovec.iov_base = arm64_reg;
+    iovec.iov_len = sizeof arm64_reg;
+    if (ptrace(PTRACE_SETREGSET, tid, NT_PRSTATUS, &iovec))
+        perror("ptrace(PTRACE_SETREGSET)");
+}
+#endif
+
+long pt_debugger_arm64::peek_reg(int reg) {
+    return arm64_reg[reg];
```

Comment:
The way this is set up makes it seem like `arm64_reg[NT_ARM_SYSTEM_CALL]` will always return the most recently-set syscall number, but that behaviour depends on how `syscall(int id)` was compiled. Don't know if that's of any value, though, but it might prevent some weird issue in the future.

Reply:
We cannot change the system call register since it's x8, but the system call overwrites x0. On x86, the system call number register is the one trashed by the return value. This is not the case on ARM.